### PR TITLE
#583 xASL_test_IntegrationTesting: Initial implementation

### PR DESCRIPTION
**Pull request by MichaelStritt**
### Linked issue

Check out issue #583

### How to test

- Create a `testConfig.json` in your ExploreASL/Testing directory
- Run: `[TestResults, CheckResults] = xASL_test_IntegrationTesting;`

### Comments

I know this probably doesn't fulfill all requirements yet, but I thought I could show you my initial implementation and how I want to extract the errors/warning in the future. The x struct includes a logging field now (implemented in #588). This includes both errors from the import and from the processing workflow. Those should be written into a `log.xlsx` file. It doesn't catch the various warnings and stuff right now, but the major things are nicely captured.

### CODE REVIEW

**Review by jan-petr — CHANGES_REQUESTED** [review]
I think we are nearly there. I agree with your edits. Those weird calls of ExploreASL, we have tested previously. The main thing that was missing is that during the Flavors testing, we were calling directly the ExloreASL submodules and functions like xASL_module_Import or xASL_bids_BIDS2Legacy. This did the job, but wasn't testing if the master script is working. I now replaced all the calls with calls to the master script ExploreASL. All should work and should be better for testing. 

One small thing is missing - it now initializes everything again upon each call of ExploreASL - can we do something about it? Probably not and we have to live with it, right? This would be also OK - more initializations are called because we want every call to be an independent call of ExploreASL, so that's the price we pay for this.

Once we solve this minor thing and once you check if all works (I have tested 1 flavor), then I will approve and I will merge. Note that I did some small restructuring of function names etc.

**Comment by MichaelStritt** [comment]
> One small thing is missing - it now initializes everything again upon each call of ExploreASL - can we do something about it? Probably not and we have to live with it, right? This would be also OK - more initializations are called because we want every call to be an independent call of ExploreASL, so that's the price we pay for this.

If we want the tests to be more independent, we can add a small script to remove the Matlab paths in between. Similar to the call in Henks old test script:

```matlab
% Remove ExploreASL paths
warning('off','MATLAB:rmpath:DirNotFound');
rmpath(genpath(x.MyPath));
warning('on','MATLAB:rmpath:DirNotFound');
```

> Once we solve this minor thing and once you check if all works (I have tested 1 flavor), then I will approve and I will merge. Note that I did some small restructuring of function names etc.

**Comment by jan-petr** [comment]
> > One small thing is missing - it now initializes everything again upon each call of ExploreASL - can we do something about it? Probably not and we have to live with it, right? This would be also OK - more initializations are called because we want every call to be an independent call of ExploreASL, so that's the price we pay for this.
> 
> If we want the tests to be more independent, we can add a small script to remove the Matlab paths in between. Similar to the call in Henks old test script:
> 
> ```matlab
> % Remove ExploreASL paths
> warning('off','MATLAB:rmpath:DirNotFound');
> rmpath(genpath(x.MyPath));
> warning('on','MATLAB:rmpath:DirNotFound');
> ```
> 
> > Once we solve this minor thing and once you check if all works (I have tested 1 flavor), then I will approve and I will merge. Note that I did some small restructuring of function names etc.

Yes - good idea. Can you please add that?

**Comment by MichaelStritt** [comment]
> 
> 
> > > One small thing is missing - it now initializes everything again upon each call of ExploreASL - can we do something about it? Probably not and we have to live with it, right? This would be also OK - more initializations are called because we want every call to be an independent call of ExploreASL, so that's the price we pay for this.
> > 
> > 
> > If we want the tests to be more independent, we can add a small script to remove the Matlab paths in between. Similar to the call in Henks old test script:
> > ```matlab
> > % Remove ExploreASL paths
> > warning('off','MATLAB:rmpath:DirNotFound');
> > rmpath(genpath(x.MyPath));
> > warning('on','MATLAB:rmpath:DirNotFound');
> > ```
> > 
> > > Once we solve this minor thing and once you check if all works (I have tested 1 flavor), then I will approve and I will merge. Note that I did some small restructuring of function names etc.
> 
> Yes - good idea. Can you please add that?

It's a bit difficult, since we have to make sure that we run it in the correct code section and we're still able to reinitialize etc.

The script to remove the paths (920fc28) is pretty simple and I found two code sections where we can easily add this (7ac0165).

**Comment by HenkMutsaerts** [comment]
Looks good, just don't store these in an `XLSX` but rather `TSV` (also according to BIDS). You can use `xASL_tsvWrite`.

File: Functions/xASL_adm_RemoveDirectories.m [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+% Copyright 2015-2021 ExploreASL
+
+    %% Check input
+    if nargin < 1
+        error('Please provide a root directory...');
+    end
+    
+    % ExploreASL
+    try
+        warning('off', 'all');
```
Comment (HenkMutsaerts) [inline Functions/xASL_adm_RemoveDirectories.m:25]: This is tricky, as it disables all warnings (so debugging becomes tricky) and enables all warnings (even those that a user might have disabled!

So this is not good practice. You can disable warnings specifically, google it or check in our current code. You run this script, then when a warning occurs, run `lastwarn` which gives you also the warning ID, and then you can disable this specific warning ID.

File: Testing/xASL_test_IntegrationTesting.m [hunk 2]
```diff
@@ -109,14 +116,30 @@
...
+    
     % Copy flavor
     fprintf('Copy %s...\n', flavorName);
     xASL_Copy(fullfile(databaseRoot,flavorName),fullfile(testingRoot,'FlavorDatabaseTest',flavorName),1);
     
     % Remove rawdata folder from testing dataset
     xASL_delete(fullfile(testingRoot,'FlavorDatabaseTest',flavorName,'rawdata'),1);
+    
+    % Create default dataPar.json
+    defaultDataPar.x.subject_regexp = '^sub-.*$';
```
Comment (HenkMutsaerts) [inline Testing/xASL_test_IntegrationTesting.m:133]: Why did you add these here, when they are defaulted by ExploreASL itself? It seems unnecessary convoluted for me now, you can probably remove many testing lines and make the scripts more efficient and readable?

**Review by HenkMutsaerts — CHANGES_REQUESTED** [review]
Very nice work! Though I wonder if not many of the things that are now stated explicitly in the testing scripts should be stated inside the actual ExploreASL scripts, and the testing iterates them, so simplification. E.g., the logging should be part of ExploreASL as well right, why only logging when testing and not when a user runs the scripts? 

I would make the actual running and testing as equal as possible, so debugging becomes the same if a report comes from our internal testing or from the user.

**Comment by MichaelStritt** [comment]
> [...] logging should be part of ExploreASL as well right, why only logging when testing and not when a user runs the scripts?

Logging is part of ExploreASL already. It can be still improved though. Check out `xASL_Iteration` lines 399-400:             

```
% Catch error in logging field
[x] = xASL_qc_AddLoggingInfo(x, ME);
```

The only thing that the integration testing script does now, is to combine all of the logs from the different flavors into an xlsx and a tsv file. IMHO we don't need this every time somebody runs ExploreASL in general.

File: Functions/xASL_adm_RemoveDirectories.m (hunk 1 above)
Comment (MichaelStritt) [reply Functions/xASL_adm_RemoveDirectories.m]: 8cdb69b

File: Testing/xASL_test_IntegrationTesting.m (hunk 2 above)
Comment (MichaelStritt) [reply Testing/xASL_test_IntegrationTesting.m:133]: We need ...
```
    defaultDataPar.x.bUseMNIasDummyStructural = 1;
    defaultDataPar.x.Quality = 0;
```
... those are not default.

Comment (MichaelStritt) [reply Testing/xASL_test_IntegrationTesting.m]: 237f78d

**Review by jan-petr — APPROVED** [review]
OK.

**Comment by HenkMutsaerts** [comment]
> > [...] logging should be part of ExploreASL as well right, why only logging when testing and not when a user runs the scripts?
> 
> Logging is part of ExploreASL already. It can be still improved though. Check out `xASL_Iteration` lines 399-400:
> 
> ```
> % Catch error in logging field
> [x] = xASL_qc_AddLoggingInfo(x, ME);
> ```
> 
> The only thing that the integration testing script does now, is to combine all of the logs from the different flavors into an xlsx and a tsv file. IMHO we don't need this every time somebody runs ExploreASL in general.

OK nice!

File: Testing/xASL_test_IntegrationTesting.m [hunk 3]
```diff
@@ -232,12 +232,19 @@
 %% Write logging information to XLSX table
 function xASL_test_writeLoggingTable(filePath,logTable)
 
-    % Write file
+    % Write XLSX file
     try
-        writetable(logTable,filePath,'WriteVariableNames',true);
+        writetable(logTable,[filePath '.xlsx'],'WriteVariableNames',true);
```
Comment (HenkMutsaerts) [inline Testing/xASL_test_IntegrationTesting.m:237]: Why is `xlsx` so useful, can we not keep TSV only? That would be better for multiple reasons: it is more open-access, easier to read, lighter weight.

File: Functions/xASL_adm_RemoveDirectories.m [hunk 4]
```diff
@@ -22,9 +22,9 @@ function xASL_adm_RemoveDirectories(root)
     
     % ExploreASL
     try
-        warning('off', 'all');
+        warning('off', 'MATLAB:rmpath:DirNotFound');
```
Comment (HenkMutsaerts) [inline Functions/xASL_adm_RemoveDirectories.m:25]: great, perhaps add some comments about what we do here, and why we use the try-catch construction for this. It seems not required? Matlab never gives an error with `rmpath`, only warnings? I would avoid try-catch constructions if we can before the code becomes too convoluted

File: Testing/xASL_test_IntegrationTesting.m [hunk 5]
```diff
@@ -130,10 +130,8 @@
     xASL_delete(fullfile(testingRoot,'FlavorDatabaseTest',flavorName,'rawdata'),1);
     
     % Create default dataPar.json
-    defaultDataPar.x.subject_regexp = '^sub-.*$';
     defaultDataPar.x.bUseMNIasDummyStructural = 1; % when no structural data, use ASL-MNI registration
```
Comment (HenkMutsaerts) [inline Testing/xASL_test_IntegrationTesting.m:135]: I thought this wasn't the case for all flavors, you mind checking if all flavors don't have an T1.nii, otherwise we should remove any structural NIfTIs (perhaps in another issue)

**Review by HenkMutsaerts — CHANGES_REQUESTED** [review]
Few minor issues but otherwise happy :)

File: Functions/xASL_adm_RemoveDirectories.m (hunk 4 above)
Comment (MichaelStritt) [reply Functions/xASL_adm_RemoveDirectories.m:25]: Yeah, but you get like a hundred warnings and I just wanted to make sure that the pipeline is not killed because of this script which is only there to improve test independence.

File: Testing/xASL_test_IntegrationTesting.m (hunk 3 above)
Comment (jan-petr) [reply Testing/xASL_test_IntegrationTesting.m:237]: Agree. TSV is easy to import to Excel. Was there some special reason @MichaelStritt ?

File: Testing/xASL_test_IntegrationTesting.m (hunk 5 above)
Comment (MichaelStritt) [reply Testing/xASL_test_IntegrationTesting.m:135]: I think this isn't the case for all flavors. Jan is the expert here though. I just wanted to create an initial testing script where you can run each flavor from start to finish with less set up and stuff like in Jans version, but his one probably works better.

File: Testing/xASL_test_IntegrationTesting.m (hunk 3 above)
Comment (MichaelStritt) [reply Testing/xASL_test_IntegrationTesting.m:237]: I don't know... I just like it because I can double click open it with excel or libre office and it's nicer too look at than a tsv text editor. This isn't part of the pipeline where we have to be 100% bids compliant. We had the exact same discussion with the previous versions a few months ago as well though, I think we just have a different opinion here.

File: Functions/xASL_adm_RemoveDirectories.m (hunk 4 above)
Comment (jan-petr) [reply Functions/xASL_adm_RemoveDirectories.m:25]: I am a bit confused:)
It is OK to switch off warnings for rmpath - makes total sense to avoid too much unnecessary warnings - so just add a comment explaining in the code why you do this.

But the question was mostly the try-catch - why is that needed. Can theoretically `warning` crash or `rmpath` crash - opinion of Henk is that these cannot really crash, or can they?

File: Testing/xASL_test_IntegrationTesting.m (hunk 5 above)
Comment (jan-petr) [reply Testing/xASL_test_IntegrationTesting.m:135]: Good point. This is not really needed. Some datasets have T1. And there were 1-2 datasets where the bUseMNIasDummy option was needed. But I have added the DataPar to the flavors with this option on. So for the purpose of testing flavors - we don't need it.

And we should probably exclude it so that it doesn't use MNI for those data that have T1.

We definitively do **not** want to remove T1.niis from flavors, because we want to test the import of them as well!!!

File: Testing/xASL_test_IntegrationTesting.m (hunk 3 above)
Comment (jan-petr) [reply Testing/xASL_test_IntegrationTesting.m:237]: ah - true. That's a test only. Fine to keep XLSX. Actually - my editor opens Excel with tsvs as well - that's why I don't care that much :D

File: Functions/xASL_adm_RemoveDirectories.m (hunk 4 above)
Comment (MichaelStritt) [reply Functions/xASL_adm_RemoveDirectories.m]: > Can theoretically warning crash or rmpath crash - opinion of Henk is that these cannot really crash, or can they?

I actually wasn't sure about that, but I thought better safe than sorry. Didn't knew you had so strong opinions on this :D

43f5305 👍

File: Testing/xASL_test_IntegrationTesting.m (hunk 5 above)
Comment (MichaelStritt) [reply Testing/xASL_test_IntegrationTesting.m]: > We definitively do not want to remove T1.niis from flavors, because we want to test the import of them as well!!!

Agree, that would make the test database worse.

> And we should probably exclude it so that it doesn't use MNI for those data that have T1.

Oh, sorry, didn't knew that you added a dataPar.json for the ones without T1.


f668525 👍

File: Testing/xASL_test_IntegrationTesting.m [hunk 6]
```diff
@@ -132,7 +132,6 @@
     xASL_delete(fullfile(testingRoot,'FlavorDatabaseTest',flavorName,'rawdata'),1);
     
     % Create default dataPar.json
-    defaultDataPar.x.bUseMNIasDummyStructural = 1; % when no structural data, use ASL-MNI registration
```
Comment (HenkMutsaerts) [inline Testing/xASL_test_IntegrationTesting.m]: This parameter seems necessary for the flavor database, which hasn't got structural images for many flavors.

**Review by HenkMutsaerts — CHANGES_REQUESTED** [review]

File: Testing/xASL_test_IntegrationTesting.m (hunk 5 above)
Comment (MichaelStritt) [reply Testing/xASL_test_IntegrationTesting.m]: > Good point. This is not really needed. Some datasets have T1. And there were 1-2 datasets where the bUseMNIasDummy option was needed. But I have added the DataPar to the flavors with this option on.

Are you sure about that? For example the first one (**GE_PCASL_2DEPI_volunteer**) does not have a T1 scan **and** there is no `dataPar.json` with the `x.bUseMNIasDummyStructural = 1;` setting.

Comment (jan-petr) [reply Testing/xASL_test_IntegrationTesting.m]: @MichaelStritt very good point :D Sorry for the confusion. I just did remember that setting this parameter wasn't necessary and that everything was running without it. But that's because we already have it in the xASL_test_Flavors.m :) I didn't realize that we need it in that other function because it is another unrelated function :D And that other parameter I was talking about is something else.

So indeed, we need to have this :) Easy solution - I will just remove your last commit and then we can merge everything. I will do that to avoid conflicts :)

File: Testing/xASL_test_IntegrationTesting.m (hunk 6 above)
Comment (jan-petr) [reply Testing/xASL_test_IntegrationTesting.m]: You are of course right. My fault of insisting that Michael should remove it... I'm reverting this change.

**Review by HenkMutsaerts — APPROVED** [review]
